### ISSUE TITLE
AndroidJavaException: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/braintreepayments/api/PayPalAccountNonce; <api:paypal:4.8.2>

### ISSUE THREAD

**Issue by nir-jackson-papaya** [braintree/braintree_android#511]
* SDK/Library version: com.braintreepayments.api:paypal:4.8.2
* Environment: Sandbox
* Braintree dependencies: com.braintreepayments.api:paypal:4.8.2

I'm having issues migrating to v4.
Our app currently works with v3.
After completing the migration I created an aar to be imported into our app (Unity).
When trying to access the api i get this error:

AndroidJavaException: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/braintreepayments/api/PayPalAccountNonce;
java.lang.NoClassDefFoundError: Failed resolution of: Lcom/braintreepayments/api/PayPalAccountNonce;
at java.lang.reflect.Executable.getParameterTypesInternal(Native Method)
at java.lang.reflect.Method.getParameterTypes(Method.java:186)
at java.lang.Class.getDeclaredMethods(Class.java:1884)
at com.unity3d.player.ReflectionHelper.getMethodID(Unknown Source:26)
at com.unity3d.player.UnityPlayer.nativeRender(Native Method)
at com.unity3d.player.UnityPlayer.access$300(Unknown Source:0)
at com.unity3d.player.UnityPlayer$e$1.handleMessage(Unknown Source:95)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:233)
at com.unity3d.player.UnityPlayer$e.run(Unknown Source:20)
Caused by: java.lang.ClassNotFoundException: com.braintreepayments.api.PayPalAccountNonce

**Comment by sshropshire** [comment]
Hi @nir-jackson-papaya thanks for using the Braintree SDK for Android. I'm not as familiar with Unity, although a quick Google search shows there may be a `minification` setting of some sort.

Are you aware of this option? If so I would seek to disable minification. From the stack trace, it appears that Unity is using reflection to access that class.

**Comment by nir-jackson-papaya** [comment]
Hi @sshropshire 
Thank you for your reply.
The issue was with our Unity version 2019.3.32. The builtin Gradle 5.1.1 and OpenJDK 1.8.0_152 were not compatible with the addition of the <queries> tag added to the AndroidManifest. Resulting in a failure to get the api library.
We upgraded to Unity 2020 and this was resolved.

I'm currently facing a different issue.
From within my UnityPlayerActivity I create a new AppCompatActivity that implements the [https://github.com/braintree/braintree_android/blob/master/v4_MIGRATION_GUIDE.md#paypal](PayPalClient).

I then call a static function that sets the required variables like token, amount and callback to my UnityPlayerActivity.
Then I create a new Intent like this:
```
public static Intent getBraintreeIntent(Activity unityActivity){
        return new Intent(unityActivity, PayPalClient.class);
    }

```
Then from within my UnityPlayerActivity I call startActivity with the created intent.
After this the app crashes and I see in the logcat this Error:

> 03-09 16:59:43.701 22223 22223 I System.out: [Unity BrainTree_Internal] onCreate 2
> 03-09 16:59:43.704 22223 32283 E sqlite3_android: [IKR-38846] ONEPLUS_NAME_PARTS_MATCH SQLITE_OK
> 03-09 16:59:43.714 22223 22223 D AndroidRuntime: Shutting down VM
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: FATAL EXCEPTION: main
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: Process: com.papaya.solitairecash, PID: 22223
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/braintreepayments/api/AnalyticsDatabase;
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at com.braintreepayments.api.AnalyticsClient.<init>(AnalyticsClient.java:53)
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at com.braintreepayments.api.BraintreeClient.createDefaultParams(BraintreeClient.java:66)
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at com.braintreepayments.api.BraintreeClient.createDefaultParams(BraintreeClient.java:38)
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at com.braintreepayments.api.BraintreeClient.<init>(BraintreeClient.java:81)
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at com.papayaandroid.braintreepaypal.BrainTreePapaya.onCreate(BrainTreePapaya.java:38)
> 03-09 16:59:43.715 22223 22223 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:8006)
> 

This is my onCreate:
` @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            System.out.println("[Unity BrainTree_Internal] onCreate " + token);
            braintreeClient = new BraintreeClient(this, token);
            System.out.println("[Unity BrainTree_Internal] onCreate braintreeClient");
            payPalClient = new PayPalClient(braintreeClient);
            System.out.println("[Unity BrainTree_Internal] onCreate payPalClient");
            payPalCheckout();
            System.out.println("[Unity BrainTree_Internal] onCreate payPalCheckout");
        }catch (Exception e){
            System.out.println("[Unity BrainTree_Internal] Error " + e);
        }
    }`
    
    For some reason the AnalyticsDatabase isn't resolved. Could you help me understand why this is? 
    
    Thanks.

**Comment by sshropshire** [comment]
Thanks @nir-jackson-papaya this seems like it could be another minification error. I'm not sure why it would have trouble finding the `AnalyticsDatabase` class. It is `package-private`, but so is `AnalyticsClient` – I would expect it to crash much sooner if that were an issue. My hunch is Unity may be applying its own proguard rules to the project that mangle the name of the class.

Would you be able to create a separate issue for this? I'm less familiar with Unity. Maybe in a new thread we could post some screenshots of the Unity/Gradle dependency process for future reference.

**Comment by nir-jackson-papaya** [comment]
Hey @sshropshire 
I was able to resolve all of my issues. The problem was with dependencies. Unity does not resolve dependencies automatically. I had to manually download and import missing dependencies into the Unity project.
For example: the reason AnalyticsDatabase wasn't resolved was because it uses androidx.room which was missing from our project. I simply downloaded the missing libraries aar's from [Google's Maven Repository](https://maven.google.com/web/index.html) until all of the dependencies were resolved.
It is working and we expect to use Braintree V4 in our next version.

Thank you
